### PR TITLE
WIP: netcdf

### PR DESCRIPTION
**Pull request by ocefpaf**
Sadly this package was renamed to `libnetcdf` in the default channel and I guess we should keep that to maintain compatibility.

This is an attempt to update the `netcdf` package to the latest version and compiled against conda-forge's `hdf5`.

### CODE REVIEW

File: recipes/libnetcdf/bld.bat [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+cmake --build .
+if errorlevel 1 exit 1
+
+cmake --build . --target INSTALL
+if errorlevel 1 exit 1
+
+:: Remove msvc redist files.
+cd %LIBRARY_BIN%
+rm msvc*.dll
```
Comment (ocefpaf) [inline recipes/libnetcdf/bld.bat]: I never tried this recipe on Windows. @rgrout can you take a look please?

File: recipes/libnetcdf/build.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    --without-ssl \
+    --without-libidn \
+    --disable-ldap \
+    --prefix=$PREFIX
+
+make
+make check
+make install
+
+rm -rf $PREFIX/share
```
Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: Don't know the reason behind this. I believe this recipe is from the `anaconda-recipes` repostory.

File: recipes/libnetcdf/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+about:
+    home: http://www.unidata.ucar.edu/software/netcdf/
+    license: MIT
+    summary: libraries and data formats that support array-oriented scientific data
+
+extra:
+    recipe-maintainers:
+        - jakirkham
+        - rgrout
```
Comment (ocefpaf) [inline recipes/libnetcdf/meta.yaml]: @jakirkham and @groutr I added you here because of `hdf5`. Let me know if that is OK for you.

File: recipes/libnetcdf/build.sh (hunk 2 above)
Comment (msarahan) [inline recipes/libnetcdf/build.sh]: Probably just saving space?  I don't know otherwise.

Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: That is my guess too.

File: recipes/libnetcdf/bld.bat [hunk 4]
```diff
@@ -0,0 +1,23 @@
+mkdir %SRC_DIR%\build
+cd %SRC_DIR%\build
+
+set HDF5_DIR=%LIBRARY_PREFIX%\cmake
+
+if "%ARCH%" == "64" (
+  set CMAKE_GENERATOR="Visual Studio 10 2010 Win64"
```
Comment (msarahan) [inline recipes/libnetcdf/bld.bat]: We need something better here to build for the different MSVC versions.  You should be able to use one of the two examples from https://github.com/conda/conda-build/wiki/Windows-recipe-patterns#using-cmake-with-visual-studio

Comment (ocefpaf) [inline recipes/libnetcdf/bld.bat]: Thanks. I will take a look at that.

File: recipes/libnetcdf/bld.bat (hunk 1 above)
Comment (msarahan) [inline recipes/libnetcdf/bld.bat]: it is weird to manually remove the dlls here.  Maybe the NetCDF install step copied them in?  If so, keep this removal.  Otherwise, this is cruft.

File: recipes/libnetcdf/build.sh (hunk 2 above)
Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: Those are `man` pages:

```shell
share/man/man1/nccopy.1
share/man/man1/ncdump.1
share/man/man1/ncgen.1
share/man/man1/ncgen3.1
share/man/man3/netcdf.3
```

I believe those are not used in conda anyway.

Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Yeah, removing sounds smart then.

File: recipes/libnetcdf/meta.yaml (hunk 3 above)
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Mainly just use `hdf5`/`h5py`, but it's fine.

File: recipes/libnetcdf/build.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+export CFLAGS="-I$PREFIX/include $CFLAGS"
+export LDFLAGS="-L$PREFIX/lib $LDFLAGS"
+export DYLD_LIBRARY_PATH=$PREFIX/lib
+
+# https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg11423.html
+export LD_LIBRARY_PATH=$PREFIX/lib
+
+./configure \
```
Comment (kmuehlbauer) [inline recipes/libnetcdf/build.sh]: Are we bound to use `configure` here? I had a look at [how unidata is testing](https://travis-ci.org/Unidata/netcdf-c/jobs/118246341). They also use a special docker image, but use cmake instead of configure.

Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: > Are we bound to use `configure` here?

Nope. We can try `cmake`

> They also use a special docker image, but use cmake instead of configure.

I will give that a try later. Thanks!

Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: So, I know @groutr [mentioned]( https://github.com/conda-forge/staged-recipes/pull/192#discussion_r57233201 ) that he tried `cmake` to build `hdf5` and had issues. The issues may also present themselves here so we should be careful. Personally, given `hdf5` had issues, I prefer `configure` at present.

File: recipes/libnetcdf/meta.yaml (hunk 3 above)
Comment (groutr) [inline recipes/libnetcdf/meta.yaml]: That's fine.  it should be `groutr` though.

File: recipes/libnetcdf/build.sh (hunk 5 above)
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Really hoped GitHub would handle that link better. Anyways, he was seeing segfaults when using `hdf5`/`h5py` with Python when `hdf5` was built with `cmake`. I prefer using `configure` to segfaults any day. If this situation is any bit similar here, then we would be wise to stick with what works.

Comment (kmuehlbauer) [inline recipes/libnetcdf/build.sh]: @jakirkham Thanks for the heads up. But it may be of use in sorting out the correct `configure` switches.

Comment (groutr) [inline recipes/libnetcdf/build.sh]: It was weird, and I'm still not sure of the root cause.  I'm curious as to why the recommend cmake, yet their official binaries appear to be still using ./configure.

File: recipes/libnetcdf/build.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+export CFLAGS="-I$PREFIX/include $CFLAGS"
+export LDFLAGS="-L$PREFIX/lib $LDFLAGS"
+export DYLD_LIBRARY_PATH=$PREFIX/lib
```
Comment (groutr) [inline recipes/libnetcdf/build.sh]: This might be problematic.  We should avoid DYLD_LIBRARY_PATH because in recent versions of OSX, they don't work unless SIP is disabled.

File: recipes/libnetcdf/meta.yaml (hunk 3 above)
Comment (ocefpaf) [inline recipes/libnetcdf/meta.yaml]: Oops! Fixing that now...

File: recipes/libnetcdf/build.sh (hunk 5 above)
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Let's get a stable release out first and then people can play with adding `cmake` if they so desire. Sound good?

File: recipes/libnetcdf/build.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+export CFLAGS="-I$PREFIX/include $CFLAGS"
+export LDFLAGS="-L$PREFIX/lib $LDFLAGS"
+
+# https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg11423.html
+export LD_LIBRARY_PATH=$PREFIX/lib:$LD_LIBRARY_PATH
```
Comment (groutr) [inline recipes/libnetcdf/build.sh]: And if this is only needed for make check, maybe it should be 
`LD_LIBRARY_PATH=$PREFIX/lib:$LD_LIBRARY_PATH make check` (set LD_LIBRARY_PATH just for make check)
I don't think it should be set during `make`

Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: Makes sense.

File: recipes/libnetcdf/bld.bat (hunk 1 above)
Comment (groutr) [inline recipes/libnetcdf/bld.bat]: These particular lines may not be needed anymore.  I can't verify until I get my windows vm repaired.

File: recipes/libnetcdf/bld.bat [hunk 8]
```diff
@@ -0,0 +1,26 @@
+mkdir %SRC_DIR%\build
+cd %SRC_DIR%\build
+
+
+set LIB=%LIBRARY_LIB%;%LIB%
+set LIBPATH=%LIBRARY_LIB%;%LIBPATH%
+set INCLUDE=%LIBRARY_INC%;%INCLUDE%
+set HDF5_DIR=%LIBRARY_PREFIX%\cmake
+
+cmake -G "NMake Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=%LIBRARY_PREFIX% -DCMAKE_INSTALL_PREFIX:PATH=%LIBRARY_PREFIX% %SRC_DIR%
```
Comment (groutr) [inline recipes/libnetcdf/bld.bat]: cmake needs to be a build dependency for windows, I think.

**Comment by dopplershift** [comment]
While doing this, we need to do something to address conda-forge/netcdf4-feedstock#4.

File: recipes/libnetcdf/bld.bat (hunk 1 above)
Comment (ocefpaf) [inline recipes/libnetcdf/bld.bat]: OK. I am removing them.

File: recipes/libnetcdf/bld.bat (hunk 8 above)
Comment (ocefpaf) [inline recipes/libnetcdf/bld.bat]: Not necessarily we do have `cmake` pre-installed on AppVeyor. Not sure what is the best way to go though.

Comment (jakirkham) [inline recipes/libnetcdf/bld.bat]: We should use our `cmake`, no? Is there some reason not too?

Comment (ocefpaf) [inline recipes/libnetcdf/bld.bat]: Done!

**Comment by ocefpaf** [comment]
Note to self. I removed homebrew from this PR as a test and got:

```shell
checking whether the C compiler works... no
configure: error: in `/opt/conda/conda-bld/work':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

Now I am testing the default channel gcc, but I am not sure that is the way we want to go here.

File: recipes/libnetcdf/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        - vc9  # [win and py27]
+        - vc10  # [win and py34]
+        - vc14  # [win and py35]
+
+requirements:
+    build:
+        - m4  # [unix]
+        - msinttypes  # [win]
+        - cmake  # [win]
+        - curl
```
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Where is `curl` used here?

File: recipes/libnetcdf/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        - hdf5
+        - zlib >=1.2.8
+    run:
+        - curl
```
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Guessing it is not needed after the build.

File: recipes/libnetcdf/build.sh [hunk 11]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+if [ $(uname) == Darwin ]; then
```
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Maybe we should add a check for [`TRAVIS`]( https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables ) too.

File: recipes/libnetcdf/build.sh [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+
+export CFLAGS="-I$PREFIX/include $CFLAGS"
+export LDFLAGS="-L$PREFIX/lib $LDFLAGS"
+
+./configure \
+    --enable-shared \
+    --enable-netcdf-4 \
+    --enable-dap \
+    --without-ssl \
```
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Would it be worth adding OpenSSL and enabling this?

File: recipes/libnetcdf/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+source:
+    git_url: https://github.com/Unidata/netcdf-c.git
+    git_tag: v{{ version }}
+
+build:
+    number: 0
+    features:
+        - vc9  # [win and py27]
+        - vc10  # [win and py34]
+        - vc14  # [win and py35]
```
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Might need `python` as a build requirement for this to work.

**Comment by jakirkham** [comment]
Might need something like [this]( https://github.com/conda-forge/vigra-feedstock/blob/master/recipe/build.sh#L10-L11 ) for Mac.

File: recipes/libnetcdf/build.sh (hunk 12 above)
Comment (kmuehlbauer) [inline recipes/libnetcdf/build.sh]: Is this a correct switch to build libnetcdf? I did not find anything about ssl at unidata (http://www.unidata.ucar.edu/software/netcdf/docs/getting_and_building_netcdf.html) Can you shed some light, @jakirkham

**Comment by pelson** [comment]
What are we reviewing here @ocefpaf. Couldn't you just rename the package in the feedstock, and then rename that feedstock?

File: recipes/libnetcdf/meta.yaml (hunk 9 above)
Comment (ocefpaf) [inline recipes/libnetcdf/meta.yaml]: netcdf depends on curl to support OPeNDAP.

File: recipes/libnetcdf/meta.yaml (hunk 10 above)
Comment (ocefpaf) [inline recipes/libnetcdf/meta.yaml]: It is.

File: recipes/libnetcdf/build.sh (hunk 11 above)
Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: This line should go. It is saver to do that in the `.CI` (and it was already done.)

File: recipes/libnetcdf/build.sh (hunk 12 above)
Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: @kmuehlbauer is correct. That was in the original recipe, but it is not valid. (Also the two options below.)

**Comment by ocefpaf** [comment]
> What are we reviewing here @ocefpaf. Couldn't you just rename the package in the feedstock, and then rename that feedstock?

We don't have a feedstock `libnetcdf` (nor `netcdf`). My name comment at the top was just a rant.

**Comment by ocefpaf** [comment]
I got Windows and Linux working, but I am clueless on OS X:

```shell
checking for gcc... gcc
checking whether the C compiler works... no
configure: error: in `/Users/travis/miniconda/conda-bld/work/netcdf-c-4.4.0':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

File: recipes/libnetcdf/build.sh [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  export LIBS="-lnetcdf -lhdf5 -lhdf5_hl"
+fi
+
+if [[ $(uname) == Darwin ]]; then
+  export LIBRARY_SEARCH_VAR=DYLD_FALLBACK_LIBRARY_PATH
+elif [[ $(uname) == 'Linux' ]]; then
+  export LIBRARY_SEARCH_VAR=LD_LIBRARY_PATH
+fi
+
+export LDFLAGS="-L$PREFIX/lib"
```
Comment (groutr) [inline recipes/libnetcdf/build.sh]: It's important to always prepend or append to LDFLAGS, CFLAGS, etc.  Conda-build does use the variable to define things in the build environment.
`export LDFLAGS=-L$PREFIX/lib $LDFLAGS`

Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: Thanks!

**Comment by groutr** [comment]
The failure on OSX seems to be due to libcurl.
```
configure:17091: checking size of off_t
4812 configure:17096: gcc -o conftest -I/Users/rgrout/miniconda/envs/_build/include  -arch x86_64 -I/Users/rgrout/miniconda/envs/_build/include  -L/Users/rgrout/miniconda/envs/_build/lib  -arch x8     6_64 conftest.c -lcurl  >&5
4813 configure:17096: $? = 0
4814 configure:17096: ./conftest
4815 dyld: Library not loaded: libcurl.4.dylib
4816   Referenced from: /Users/rgrout/miniconda/conda-bld/work/netcdf-c-4.4.0/./conftest
4817   Reason: Incompatible library version: conftest requires version 9.0.0 or later, but libcurl.4.dylib provides version 7.0.0
4818 ./configure: line 1976: 84607 Trace/BPT trap: 5       ./conftest$ac_exeext
```
Not sure if it's picking up the system curl or not, but the conda version of curl is the correct version
```
         name libcurl.4.dylib (offset 24)
   time stamp 1 Wed Dec 31 18:00:01 1969
      current version 9.0.0
compatibility version 9.0.0
```

**Comment by ocefpaf** [comment]
I don't see that failure on our OS X image at `Travis-CI`. See https://github.com/conda-forge/staged-recipes/pull/213#issuecomment-205770942

File: recipes/libnetcdf/meta.yaml (hunk 13 above)
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Fixed below.

**Comment by groutr** [comment]
I commented out the first if-block.  For some reason that messes up configure.

**Comment by groutr** [comment]
What I posted in https://github.com/conda-forge/staged-recipes/pull/213#issuecomment-206039528 was related to https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg12897.html.  Were you getting the error loading the netcdf shared library?

**Comment by ocefpaf** [comment]
> https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg12897.html. Were you getting the error loading the netcdf shared library?

Nope. I don't have a Mac to test it locally. All I got is the `Travis-CI OS X` logs.

**Comment by groutr** [comment]
Is curl being provided by the defaults repo, or does conda-forge have its own version of curl?

**Comment by ocefpaf** [comment]
I just disabled `OPeNDAP` to avoid possible `curl` issues and I found out that we have trouble with `hdf5` (and probably `zlib`) too:

```shell
libtool: link: gcc -I/Users/travis/miniconda/envs/_build/include -arch x86_64 -arch x86_64 -o .libs/ncgen generate.o main.o cdata.o bindata.o genchar.o cvt.o data.o debug.o escapes.o genc.o genbin.o generr.o genlib.o getfill.o odom.o offsets.o semantics.o dump.o util.o bytebuffer.o list.o genf77.o f77data.o genj.o jdata.o nc_iter.o ConvertUTF.o ncgeny.o  -L/Users/travis/miniconda/envs/_build/lib ../liblib/.libs/libnetcdf.dylib /Users/travis/miniconda/envs/_build/lib/libhdf5_hl.dylib /Users/travis/miniconda/envs/_build/lib/libhdf5.dylib -ldl -lm -lz
Making all in ncdump
../ncgen/ncgen -lc -o ctest0.nc ../ncgen/c0.cdl > ../ncdump/ctest.c
dyld: Library not loaded: @rpath/libhdf5_hl.10.dylib
  Referenced from: /Users/travis/miniconda/conda-bld/work/netcdf-c-4.4.0/ncgen/.libs/ncgen
  Reason: image not found
/bin/sh: line 1:  8864 Trace/BPT trap: 5       ../ncgen/ncgen -lc -o ctest0.nc ../ncgen/c0.cdl > ../ncdump/ctest.c
make[2]: *** [ctest.c] Error 133
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

@jakirkham any ideas if we need to go back `hdf5` (maybe `zlib`)?

**Comment by ocefpaf** [comment]
> Is curl being provided by the defaults repo, or does conda-forge have its own version of curl?

`defaults`

**Comment by ocefpaf** [comment]
Removed `hdf5` (`zlib`) and `curl` from `libnetcdf` on OS X and things work. That is a no go! But I don't know how to proceed here.

**Comment by pelson** [comment]
@ocefpaf - anything I can do to help here?

**Comment by ocefpaf** [comment]
> @ocefpaf - anything I can do to help here?

@pelson I cannot get `netcdf` with netcdf4 support to build on OS X. That usually means a problem liking to `hdf5`. I can build `netcdf` with `netcdf3` support though.

I removed that workaround so you can see the failure.

PS: Maybe we should get the `hdf4` sorted first before merging this, so we can add `hdf4` support too. (That might require a `zlib` package.)

**Comment by jakirkham** [comment]
Can you refresh my memory as to what is wrong here beyond Mac doesn't build? For example, what things were disabled that need to be re-enabled?

**Comment by ocefpaf** [comment]
> Can you refresh my memory as to what is wrong here beyond Mac doesn't build?

It stops in the configure step with:

```shell
checking size of off_t... configure: error: in `/Users/travis/miniconda/conda-bld/work/netcdf-c-4.4.0':
configure: error: cannot compute sizeof (off_t)
See `config.log' for more details
Command failed: /bin/bash -x -e /Users/travis/build/conda-forge/staged-recipes/recipes/libnetcdf/build.sh
```

> For example, what things were disabled that need to be re-enabled?

I already re-enabled everything. So the recipes is the way it "should" be. I am re-starting the CIs as we now have `zlib` and `hdf4` to compile against.

**Comment by ocefpaf** [comment]
It looks like building with `hdf4` support will be another challenge. I will remove that and try again in the feedstock. Let's try to bring this home by fixing OS X first.

**Comment by jakirkham** [comment]
Weird. I didn't change the pipeline, but it thinks I did.

**Comment by jakirkham** [comment]
Anyways something is wrong with `libcurl`. Could we try removing?

**Comment by ocefpaf** [comment]
> Weird. I didn't change the pipeline, but it thinks I did.

Auto-bot-ghosts?

**Comment by ocefpaf** [comment]
> Anyways something is wrong with `libcurl`. Could we try removing?

We can't remove it as a netcdf without OPeNDAP support is useless. We probably need to fix whatever is wrong with `libcurl` :unamused: 

PS: But yes. Some tests I made above removing OPeNDAP support **and** `hdf5` support (which is `netcdf 4`) got `libnetcdf` building. However, a `netcdf 3` only without OPeNDAP is no good.

**Comment by jakirkham** [comment]
> We can't remove it as a netcdf without OPeNDAP support is useless. We probably need to fix whatever is wrong with `libcurl` :unamused:

Yeah, sorry, it's in the log. There is a complaint about missing symbols.Also, it looks like `libcurl.dylib` is using `@loader_path` for linking its dependencies.

Do we already have `curl` here? If so, we can tweak it. If not, I think it is time to add it.

**Comment by jakirkham** [comment]
I'll add it if it is not here.

**Comment by jakirkham** [comment]
Yep, it's WIP ( https://github.com/conda-forge/staged-recipes/pull/228 ). Let's take a look and see what we need to do in order to get it in here.

**Comment by ocefpaf** [comment]
> Yeah, sorry, it's in the log.

This log is long and messy. No need to be sorry.

> There is a complaint about missing symbols.Also, it looks like `libcurl.dylib` is using `@loader_path` for linking its dependencies.

OK. Let's get #228 going then.

**Comment by jakirkham** [comment]
> This log is long and messy. No need to be sorry.

Sorry, not that log. The `config.log` for spill over. So, we wouldn't have seen this on Travis CI.

> OK. Let's get #228 going then.

Yeah, I'm going to take it over. That was always the plan.

**Comment by jakirkham** [comment]
Working on it in this PR ( https://github.com/conda-forge/staged-recipes/pull/378 ).

File: recipes/libnetcdf/build.sh [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+./configure --prefix=$PREFIX \
+            --enable-shared \
+            --enable-static \
+            --enable-extra-example-tests \
+            --disable-doxygen \
+            --enable-netcdf-4 \
+            --enable-ncgen4 \
+            --enable-dap \
+            --enable-hdf4
```
Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: I finally got `hdf4` support working!
(See https://github.com/conda-forge/hdf4-feedstock/pull/2)

File: recipes/libnetcdf/build.sh [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+make
+# https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg11423.html
+eval ${LIBRARY_SEARCH_VAR}=$PREFIX/lib make check
```
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Could we please add `|| cat ncdump/test-suite.log` at the end of the `make check` line? This will let us see what is going on with the test failures.

File: recipes/libnetcdf/meta.yaml [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+requirements:
+    build:
+        - python  # [win]
+        - m4  # [unix]
+        - pkg-config  # [unix]
+        - msinttypes  # [win]
+        - cmake  # [win]
+        - curl
+        - zlib
```
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Think we need to pin `zlib` to `1.2*`. Here.

File: recipes/libnetcdf/meta.yaml [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+        - hdf4
+        - hdf5
+    run:
+        - curl
+        - zlib
```
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Also, think we need to pin `zlib` to `1.2*` here.

File: recipes/libnetcdf/meta.yaml [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+    build:
+        - python  # [win]
+        - m4  # [unix]
+        - pkg-config  # [unix]
+        - msinttypes  # [win]
+        - cmake  # [win]
+        - curl
+        - zlib
+        - hdf4
+        - hdf5
```
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Also, need `hdf5 1.8.15*`.

File: recipes/libnetcdf/meta.yaml [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    run:
+        - curl
+        - zlib
+        - hdf4
+        - hdf5
```
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Need `hdf5 1.8.15*` here too.

File: recipes/libnetcdf/meta.yaml [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+requirements:
+    build:
+        - python  # [win]
+        - m4  # [unix]
+        - pkg-config  # [unix]
+        - msinttypes  # [win]
+        - cmake  # [win]
+        - curl
+        - zlib
+        - hdf4
```
Comment (jakirkham) [inline recipes/libnetcdf/meta.yaml]: Should we be pinning `hdf4`?

**Comment by ocefpaf** [comment]
Windows and Linux are OK again, but OS X is still failing.

[This](https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg12897.html) suggests adding `LIBS="-lnetcdf -lhdf5 -lhdf5_hl"` to fix the previous failure (`configure: error: cannot compute sizeof (off_t)`), but then we get:

```shell
checking for gcc... gcc
checking whether the C compiler works... no
configure: error: in `/Users/travis/miniconda/conda-bld/work/netcdf-c-4.4.0':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

Not sure how to proceed.

**Comment by dopplershift** [comment]
Those commands were in reference to building the FORTRAN library...which is separate from the C-library. There's no way this can work--`LIBS="-lnetcdf -lhdf5 -lhdf5_hl"` is asking it to link against `libnetcdf.so` **before** we've even built it.

**Comment by dopplershift** [comment]
So `off_t` is a red-herring (thanks @WardF). The real error is:
```
configure:10231: ./conftest
dyld: Library not loaded: @rpath/libcurl.4.dylib
  Referenced from: /Users/rmay/miniconda3/conda-bld/work/netcdf-c-4.4.0/./conftest
  Reason: Incompatible library version: conftest requires version 9.0.0 or later, but libcurl.4.dylib provides version 7.0.0
./configure: line 1976: 98178 Trace/BPT trap: 5       ./conftest$ac_exeext
configure:10231: $? = 133
configure: program exited with status 133
```
The problem comes from the fact that it builds and links against conda's copy of libcurl (from conda-forge), but for some reason the config test is running against the system copy, which has the same name `libcurl.4.dylib` (at least on El Cap), but only goes up to compatibility version 7.0.0 (whatever the hell that means).

**Comment by dopplershift** [comment]
So trying to use non-system libcurl sounds like hell on OSX. I can get further by specifying `[not osx]` in the `meta.yml`. I'm currently working around problems with System Integrity Protection on El Cap causing `DYLD_LIBRARY_PATH` to be ignored by shell scripts (https://developer.apple.com/library/mac/documentation/Security/Conceptual/System_Integrity_Protection_Guide/RuntimeProtections/RuntimeProtections.html#//apple_ref/doc/uid/TP40016462-CH3-SW1); maybe Travis will work though?

**Comment by dopplershift** [comment]
So how about we just use cmake and eliminate autotools? Because this works for `build.sh` on my OSX system:
```sh
#!/bin/bash

cmake -DENABLE_DAP=ON -DCMAKE_INSTALL_PREFIX=$PREFIX .
make
make install
```
and I can even get it to link against the conda-forge libcurl. I can also say that @WardF (the netCDF maintainer) uses cmake for building. With this script and temporarily removing the hdf4 dependency due to conda-forge/hdf4-feedstock#5, I can get a usable set of netCDF4 binaries.

File: recipes/libnetcdf/build.sh (hunk 15 above)
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Please add `|| { cat config.log; exit 1; }`

File: recipes/libnetcdf/build.sh [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+            --enable-extra-example-tests \
+            --disable-doxygen \
+            --enable-netcdf-4 \
+            --enable-ncgen4 \
+            --enable-dap \
+            --enable-hdf4
+
+make
+# https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg11423.html
+eval ${LIBRARY_SEARCH_VAR}=$PREFIX/lib make check || cat ncdump/test-suite.log
```
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Also, I made a mistake here. Please update this to `|| { cat ncdump/test-suite.log; exit 1; }`.

**Comment by ocefpaf** [comment]
> Those commands were in reference to building the FORTRAN library...which is separate from the C-library. There's no way this can work--`LIBS="-lnetcdf -lhdf5 -lhdf5_hl"` is asking it to link against `libnetcdf.so` **before** we've even built it.

Yeah. That was some copy-and-pasta. I should not send PRs while attending a conference :stuck_out_tongue_closed_eyes:

**Comment by ocefpaf** [comment]
> So how about we just use cmake and eliminate autotools?

@kmuehlbauer mentioned that at some point. (Sorry, but I cannot recall. We can give it a shot tomorrow.)

> With this script and temporarily removing the hdf4 dependency due to conda-forge/hdf4-feedstock#5, I can get a usable set of netCDF4 binaries.

I have been thinking about that about that and I don't have a proper solution yet. We can:

  1. eliminate `hdf4` `ncdump`/`ncgen` in the `hdf4` recipe;
  2. clobber them in the `libnetcdf` recipe;
  3. avoid using `hdf4` in the `libnetcdf` recipe;
  4. rename  `hdf4` version of those files.

I will try `2` tomorrow and see how it goes. If anyone have a better idea please let me know.

**Comment by kmuehlbauer** [comment]
@ocefpaf :grin: I had almost forgotten that myself. It was at the beginning of this PR.
Unidata is also using some docker image and cmake to build the libnetcdf in github.com work flow. So this is worth a shot. And we surely can get advice from @WardF, if there will be some problems.

**Comment by dopplershift** [comment]
I don't think conda will let us clobber--my built netcdf4 package didn't have ncdump until I removed the conflict.

**Comment by ocefpaf** [comment]
I would not rely on conda to clobber as conda does the right thing. I would
manually substitute one by the other in the recipe. (Actually the right
thing would be to raise an error ;-)
Em 21/05/2016 00:47, "Ryan May" <notifications@github.com> escreveu:

> I don't think conda will let us clobber--my built netcdf4 package didn't
> have ncdump until I removed the conflict.
>

**Comment by ocefpaf** [comment]
Testing cmake. Everything works locally but I do not know how to build static libs and shared with it. The docs mention that I should `BUILD_SHARED_LIBS=OFF` to get the static libs, but then we won't get the shared.

Also, there are a few additional options we may use. I need `DAP` and `HDF4`, but I have no idea if it is a good idea to enable the rest (`JNA`, `MMap`, `PNetCDF`, and `NC-4 Parallel`). @dopplershift do you need/recommend testing any of these features?

```
# Features
--------
NetCDF-2 API:           yes
HDF4 Support:           yes
NetCDF-4 API:           yes
NC-4 Parallel Support:  no
PNetCDF Support:        no
DAP Support:            yes
Diskless Support:       yes
MMap Support:           no
JNA Support:            no
```

**Comment by jakirkham** [comment]
> Testing cmake. Everything works locally but I do not know how to build static libs and shared with it. The docs mention that I should `BUILD_SHARED_LIBS=OFF` to get the static libs, but then we won't get the shared.

That's fine do 2 loops for building. Once with `BUILD_SHARED_LIBS=OFF` and once with `BUILD_SHARED_LIBS=ON`.

**Comment by ocefpaf** [comment]
> That's fine do 2 loops for building.

I thought about it, but sounds "wrong." Anyways, it is working now for Linux and OS X :tada: 

But Windows, that was working before, is failing now. Not sure that is going on. We can skip Windows and open an issue on the feedstock to solve it later.

File: recipes/libnetcdf/build.sh [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+      -D ENABLE_NETCDF_4=ON \
+      -D BUILD_SHARED_LIBS=OFF \
+      -D ENABLE_TESTS=ON \
+      -D BUILD_UTILITIES=ON \
+      -D ENABLE_DOXYGEN=OFF \
+      $SRC_DIR
+make
+# ctest  # Save some time.
+make install
+
```
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: I'd add a `make clean` here. Maybe even a `make distclean` if they have one.

**Comment by jakirkham** [comment]
> > That's fine do 2 loops for building.

> I thought about it, but it sounded "wrong." Anyways... That worked for Linux and OS X :tada: 

We do it for `ncurses` and it works great. We should probably start doing it for `fftw` where this is actually recommended (instead we have copy-pasta).

> That worked for Linux and OS X :tada: 

🎉 🎉 🎉

> But Windows, that was working before, is failing now. Not sure what is going on.

> (We can skip Windows and open an issue on the feedstock to solve it later.)

Not sure either.

File: recipes/libnetcdf/build.sh (hunk 23 above)
Comment (ocefpaf) [inline recipes/libnetcdf/build.sh]: Done! (No `make distclean` btw.)

**Comment by ocefpaf** [comment]
@jakirkham if you are OK with this let's merge and iterate in the feedstock. This PR is getting hard to follow.

File: recipes/libnetcdf/build.sh (hunk 23 above)
Comment (jakirkham) [inline recipes/libnetcdf/build.sh]: Thanks. Ah, ok.

**Comment by jakirkham** [comment]
LGTM. Thanks @ocefpaf.

**Comment by jakirkham** [comment]
lol I should have renamed before merging. 😆 Oh well.

**Comment by ocefpaf** [comment]
Thanks @jakirkham!
